core/client/src/main/java/alluxio/client/block/stream/NettyPacketReader.java

```diff
@@ -0,0 +1,400 @@
+import alluxio.proto.dataserver.Protocol;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.protobuf.MessageLite;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelPipeline;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A netty packet reader that streams a region from a netty data server.
+ *
+ * Protocol:
+ * 1. The client sends a read request (id, offset, length).
+ * 2. Once the server receives the request, it streams packets to the client. The streaming pauses
+ *    if the server's buffer is full and resumes if the buffer is not full.
+ * 3. The client reads packets from the stream. Reading pauses if the client buffer is full and
+ *    resumes if the buffer is not full. If the client can keep up with network speed, the buffer
+ *    should have at most one packet.
+ * 4. The client stops reading if it receives an empty packet which signifies the end of the stream.
+ * 5. The client can cancel the read request at anytime. The cancel request is ignored by the
+ *    server if everything has been sent to channel.
+ * 6. If the client wants to reuse the channel, the client must read all the packets in the channel
+ *    before releasing the channel to the channel pool.
+ * 7. To make it simple to handle errors, the channel is closed if any error occurs.
+ */
+@NotThreadSafe
+public final class NettyPacketReader implements PacketReader {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private static final boolean CANCEL_ENABLED =
+      Configuration.getBoolean(PropertyKey.USER_NETWORK_NETTY_READER_CANCEL_ENABLED);
+  private static final int MAX_PACKETS_IN_FLIGHT =
+      Configuration.getInt(PropertyKey.USER_NETWORK_NETTY_READER_BUFFER_SIZE_PACKETS);
+  private static final long READ_TIMEOUT_MS =
+      Configuration.getLong(PropertyKey.USER_NETWORK_NETTY_TIMEOUT_MS);
+
+  private final Channel mChannel;
+  private final Protocol.RequestType mRequestType;
+  private final InetSocketAddress mAddress;
+  private final long mId;
+  private final long mStart;
+  private final long mBytesToRead;
+
+  private final ReentrantLock mLock = new ReentrantLock();
+  @GuardedBy("mLock")
+  private final Queue<ByteBuf> mPackets = new LinkedList<>();
+  @GuardedBy("mLock")
+  private Throwable mPacketReaderException = null;
+  private final Condition mNotEmptyOrFail = mLock.newCondition();
```

Comment:
The `OrFail` part of the name is a bit misleading, perhaps just `mPacketAvailable`?